lib/modules/platform/github/index.ts

```diff
@@ -671,37 +593,86 @@ function matchesState(state: string, desiredState: string): boolean {
+    config.ignorePrAuthor ||
+    (ghPr?.user?.login && config?.renovateUsername
+      ? ghPr.user.login === config.renovateUsername
+      : true)
+  );
+}
+
 export async function getPrList(): Promise<Pr[]> {
-  logger.trace('getPrList()');
-  if (!config.prList) {
-    logger.debug('Retrieving PR list');
-    let prList: GhRestPr[];
+  if (!config.prCacheReady) {
     try {
-      prList = (
-        await githubApi.getJson<GhRestPr[]>(
-          `repos/${
-            config.parentRepo || config.repository
-          }/pulls?per_page=100&state=all`,
-          { paginate: true }
-        )
-      ).body;
+      let requestsTotal = 0;
+      let apiQuotaAffected = false;
+      let needNextPageFetch = true;
+      let needNextPageSync = true;
+
+      let pageIdx = 1;
+      while (needNextPageFetch && needNextPageSync) {
+        const repo = config.parentRepo || config.repository;
+        const urlPath = `repos/${repo}/pulls?per_page=100&state=all&sort=updated&direction=desc&page=${pageIdx}`;
+
+        const opts: GithubHttpOptions = { paginate: false };
+        if (pageIdx === 1) {
+          const etag = config.prCache.etag;
+          if (etag) {
+            opts.headers = { 'If-None-Match': etag };
+          } else {
+            // Speed up initial fetch
+            opts.paginate = true;
+          }
+        }
+
+        const res = await githubApi.getJson<GhRestPr[]>(urlPath, opts);
+        apiQuotaAffected = true;
+        requestsTotal += 1;
+
+        const nextPage = parseLinkHeader(res.headers?.link)?.next;
+        needNextPageFetch = !!nextPage;
+
+        if (pageIdx === 1) {
+          if (res.statusCode === 304) {
+            apiQuotaAffected = false;
+            break;
+          }
+
+          const etag = res.headers.etag;
+          if (etag) {
+            config.prCache.etag = etag;
+          }
+
+          needNextPageFetch &&= !opts.paginate;
+        }
+
+        const { body: page } = res;
+        const renovatePrs = page.filter(isRenovateRestPr);
+        needNextPageSync = config.prCache.reconcile(renovatePrs);
+
+        pageIdx += 1;
+      }
+
+      logger.debug(
+        {
+          pullsTotal: config.prCache.getItems().length,
+          requestsTotal,
+          apiQuotaAffected,
+        },
+        `getPrList success`
+      );
     } catch (err) /* istanbul ignore next */ {
       logger.debug({ err }, 'getPrList err');
       throw new ExternalHostError(err, PlatformId.Github);
     }
-    config.prList = prList
-      .filter(
-        (pr) =>
-          config.forkMode ||
-          config.ignorePrAuthor ||
-          (pr?.user?.login && config?.renovateUsername
-            ? pr.user.login === config.renovateUsername
-            : true)
-      )
-      .map(coerceRestPr);
-    logger.debug(`Retrieved ${config.prList.length} Pull Requests`);
+
+    config.prCacheReady = true;
   }
-  return config.prList;
+
+  const ghPrs = config.prCache.getItems();
+  return ghPrs.map(coerceRestPr);
```

Comment:
I know that storing in raw format was my idea, but I'm also wondering how many times we will end up coercing the entire list in a typical run and if we should temporarily cache the coerced result?

Reply:
- Will be solved here: #15010

Reply:
what's the benefit to store in raw? the drawback is a lot more disk space and memory usage. 🤔

Reply:
Because in future we may change which fields we want to extract from it and save ourselves from needing to refetch the whole list.

Reply:
We're about to use non-branch ref (i.e. `refs/renovate/***`) in order to store renovate cache remotely. Given that, the benefit is to be able to change/refactor our PR interface without cache resets/migrations.

Reply:
Yeah, if we add, remove or rename any field in our internal PR objects then we'd need to invalidate every single pull cache of repos.

Reply:
ok, sounds reasonable.